voltdb/src/main/java/com/yahoo/ycsb/db/voltdb/VoltClient4.java

```diff
@@ -0,0 +1,315 @@
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+/*
+ * This client provideds a wrapper layer for running the Yahoo Cloud Serving
+ * Benchmark (YCSB) against VoltDB. This benchmark runs a synchronous client
+ * with a mix of the operations provided below. YCSB is open-source, and may
+ * be found at https://github.com/brianfrankcooper/YCSB. The YCSB jar must be
+ * in your classpath to compile this client.
+ */
+package com.yahoo.ycsb.db.voltdb;
+
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.Vector;
+
+import org.voltdb.VoltTable;
+import org.voltdb.client.Client;
+import org.voltdb.client.ClientResponse;
+import org.voltdb.client.ClientResponseWithPartitionKey;
+import org.voltdb.client.ProcCallException;
+import org.voltdb.sortedvolttable.VoltDBTableSortedMergeWrangler;
+
+import com.yahoo.ycsb.ByteArrayByteIterator;
+import com.yahoo.ycsb.ByteIterator;
+import com.yahoo.ycsb.DB;
+import com.yahoo.ycsb.DBException;
+import com.yahoo.ycsb.Status;
+
+/**
+ * A client that can be used by YCSB to work with VoltDB.
+ */
+public class VoltClient4 extends DB {
+
+  private static final String PROCEDURE_GET_WAS_NOT_FOUND = "Procedure Get was not found";
+  private Client mclient;
+  private byte[] mworkingData;
+  private ByteBuffer mwriteBuf;
+  private boolean useScanAll = false;
+
+  private static final Charset UTF8 = Charset.forName("UTF-8");
+
+  @Override
+  public void init() throws DBException {
+    Properties props = getProperties();
+    String servers = props.getProperty("voltdb.servers", "localhost");
+    String user = props.getProperty("voltdb.user", "");
+    String password = props.getProperty("voltdb.password", "");
+    String strLimit = props.getProperty("voltdb.ratelimit");
+    String useScanAllParam = props.getProperty("voltdb.scanall", "no");
+    
+    if (useScanAllParam.equalsIgnoreCase("YES")) { 
+      useScanAll = true;
+    }
+    
+    int ratelimit = strLimit != null ? Integer.parseInt(strLimit) : Integer.MAX_VALUE;
+    try {
+      mclient = ConnectionHelper.createConnection(Thread.currentThread().getId(), servers, user, password, ratelimit);
+
+      createYCSBDBIfNeeded();
+
+    } catch (Exception e) {
+      e.printStackTrace();
+      throw new DBException(e.getMessage());
+    }
+    mworkingData = new byte[1024 * 1024];
+    mwriteBuf = ByteBuffer.wrap(mworkingData);
+  }
+
+  /**
+   * Create the VoltDB YCSB DB if needed...
+   * 
+   * @throws Exception
+   */
+  private void createYCSBDBIfNeeded() throws Exception {
+
+    final String testString = "Test";
+
+    try {
+      ClientResponse response = mclient.callProcedure("Get", testString.getBytes(UTF8), testString);
```

Comment:
Is this going to behave properly under heavy contention (e.g. when we create 50 clients in parallel?)

Reply:
On close inspection the answer turned out to be 'no' :).  That bit has been rewritten so it behaves and tested with 100 threads.